File: ALODWebUtility/Providers/LodSiteMapProvider.vb
Namespace Providers

    Public Class LodSiteMapProvider
        Inherits System.Web.XmlSiteMapProvider

        Public Overrides Function IsAccessibleToUser(ByVal context As System.Web.HttpContext, ByVal node As System.Web.SiteMapNode) As Boolean

            If (node Is Nothing) Then
                Throw New System.ArgumentNullException("node")
            End If

            If (context Is Nothing) Then
                Throw New System.ArgumentNullException("context")
            End If

            If (Not Me.SecurityTrimmingEnabled) Then
                Return True
            End If

            Dim nodeRoles As IList = node.Roles
            Dim user As System.Security.Principal.IPrincipal = context.User

            If (user Is Nothing) OrElse (nodeRoles Is Nothing) OrElse (nodeRoles.Count = 0) Then
                Return True
            End If

            If (nodeRoles.Contains("*")) Then
                Return True
            End If

            For Each role As String In nodeRoles
                If (user.IsInRole(role)) Then
                    Return True
                End If
            Next

            Return False

        End Function

    End Class

End Namespace

File: ALODWebUtility.CS/Providers/LodSiteMapProvider.cs
using System;
using System.Collections;
using System.Web;

namespace ALODWebUtility.Providers
{
    public class LodSiteMapProvider : System.Web.XmlSiteMapProvider
    {
        public override bool IsAccessibleToUser(HttpContext context, SiteMapNode node)
        {
            if (node == null)
            {
                throw new ArgumentNullException("node");
            }

            if (context == null)
            {
                throw new ArgumentNullException("context");
            }

            if (!this.SecurityTrimmingEnabled)
            {
                return true;
            }

            IList nodeRoles = node.Roles;
            System.Security.Principal.IPrincipal user = context.User;

            if (user == null || nodeRoles == null || nodeRoles.Count == 0)
            {
                return true;
            }

            if (nodeRoles.Contains("*"))
            {
                return true;
            }

            foreach (string role in nodeRoles)
            {
                if (user.IsInRole(role))
                {
                    return true;
                }
            }

            return false;
        }
    }
}
